// ng2-iq-select2/src/app/iq-select2/iq-select2.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var iq_select2_results_component_1 = require('../iq-select2-results/iq-select2-results.component');
var forms_1 = require('@angular/forms');
var Rx_1 = require('rxjs/Rx');
var messages_1 = require('./messages');
var KEY_CODE_DOWN_ARROW = 40;
var KEY_CODE_UP_ARROW = 38;
var KEY_CODE_ENTER = 13;
var KEY_CODE_TAB = 9;
var KEY_CODE_DELETE = 8;
var VALUE_ACCESSOR = {
    provide: forms_1.NG_VALUE_ACCESSOR,
    useExisting: core_1.forwardRef(function () { return IqSelect2Component; }),
    multi: true
};
var noop = function () {
};
var IqSelect2Component = (function () {
    function IqSelect2Component() {
        this.MORE_RESULTS_MSG = 'Showing ' + messages_1.Messages.PARTIAL_COUNT_VAR + ' of ' + messages_1.Messages.TOTAL_COUNT_VAR + ' results. Refine your search to show more results.';
        this.NO_RESULTS_MSG = 'No results available';
        this.referenceMode = 'id';
        this.multiple = false;
        this.searchDelay = 250;
        this.placeholder = '';
        this.minimumInputLength = 2;
        this.disabled = false;
        this.searchIcon = 'caret';
        this.deleteIcon = 'glyphicon glyphicon-remove';
        this.messages = {
            moreResultsAvailableMsg: this.MORE_RESULTS_MSG,
            noResultsAvailableMsg: this.NO_RESULTS_MSG
        };
        this.clientMode = false;
        this.onSelect = new core_1.EventEmitter();
        this.onRemove = new core_1.EventEmitter();
        this.term = new forms_1.FormControl();
        this.resultsVisible = false;
        this.selectedItems = [];
        this.searchFocused = false;
        this.placeholderSelected = '';
        this.onTouchedCallback = noop;
        this.onChangeCallback = noop;
    }
    IqSelect2Component.prototype.ngAfterViewInit = function () {
        this.subscribeToChangesAndLoadDataFromObservable();
    };
    IqSelect2Component.prototype.subscribeToChangesAndLoadDataFromObservable = function () {
        var observable = this.term.valueChanges
            .debounceTime(this.searchDelay)
            .distinctUntilChanged();
        this.subscribeToResults(observable);
    };
    IqSelect2Component.prototype.subscribeToResults = function (observable) {
        var _this = this;
        observable
            .do(function () { return _this.resultsVisible = false; })
            .filter(function (term) { return term.length >= _this.minimumInputLength; })
            .switchMap(function (term) { return _this.loadDataFromObservable(term); })
            .map(function (items) { return items.filter(function (item) { return !(_this.multiple && _this.alreadySelected(item)); }); })
            .do(function () { return _this.resultsVisible = _this.searchFocused; })
            .subscribe(function (items) { return _this.listData = items; });
    };
    IqSelect2Component.prototype.loadDataFromObservable = function (term) {
        return this.clientMode ? this.fetchAndfilterLocalData(term) : this.fetchData(term);
    };
    IqSelect2Component.prototype.fetchAndfilterLocalData = function (term) {
        var _this = this;
        if (!this.fullDataList) {
            return this.fetchData('')
                .flatMap(function (items) {
                _this.fullDataList = items;
                return _this.filterLocalData(term);
            });
        }
        else {
            return this.filterLocalData(term);
        }
    };
    IqSelect2Component.prototype.filterLocalData = function (term) {
        var _this = this;
        return Rx_1.Observable.of(this.fullDataList.filter(function (item) { return _this.containsText(item, term); }));
    };
    IqSelect2Component.prototype.containsText = function (item, term) {
        return item.text.toUpperCase().indexOf(term.toUpperCase()) !== -1;
    };
    IqSelect2Component.prototype.fetchData = function (term) {
        var _this = this;
        return this
            .dataSourceProvider(term)
            .map(function (items) { return _this.adaptItems(items); });
    };
    IqSelect2Component.prototype.adaptItems = function (items) {
        var _this = this;
        var convertedItems = [];
        items.map(function (item) { return _this.iqSelect2ItemAdapter(item); })
            .forEach(function (iqSelect2Item) { return convertedItems.push(iqSelect2Item); });
        return convertedItems;
    };
    IqSelect2Component.prototype.writeValue = function (selectedValues) {
        if (selectedValues) {
            if (this.referenceMode === 'id') {
                this.populateItemsFromIds(selectedValues);
            }
            else {
                this.populateItemsFromEntities(selectedValues);
            }
        }
        else {
            this.selectedItems = [];
        }
    };
    IqSelect2Component.prototype.populateItemsFromEntities = function (selectedValues) {
        if (this.multiple) {
            this.handleMultipleWithEntities(selectedValues);
        }
        else {
            var iqSelect2Item = this.iqSelect2ItemAdapter(selectedValues);
            this.selectedItems = [iqSelect2Item];
            this.placeholderSelected = iqSelect2Item.text;
        }
    };
    IqSelect2Component.prototype.handleMultipleWithEntities = function (selectedValues) {
        var _this = this;
        this.selectedItems = [];
        selectedValues.forEach(function (entity) {
            var item = _this.iqSelect2ItemAdapter(entity);
            var ids = _this.getSelectedIds();
            if (ids.indexOf(item.id) === -1) {
                _this.selectedItems.push(item);
            }
        });
    };
    IqSelect2Component.prototype.populateItemsFromIds = function (selectedValues) {
        if (this.multiple) {
            this.handleMultipleWithIds(selectedValues);
        }
        else {
            this.handleSingleWithId(selectedValues);
        }
    };
    IqSelect2Component.prototype.handleMultipleWithIds = function (selectedValues) {
        var _this = this;
        if (selectedValues !== undefined && this.selectedProvider !== undefined) {
            var uniqueIds_1 = [];
            selectedValues.forEach(function (id) {
                if (uniqueIds_1.indexOf(id) === -1) {
                    uniqueIds_1.push(id);
                }
            });
            this.selectedProvider(uniqueIds_1).subscribe(function (items) {
                _this.selectedItems = items.map(_this.iqSelect2ItemAdapter);
            });
        }
    };
    IqSelect2Component.prototype.handleSingleWithId = function (id) {
        var _this = this;
        if (id !== undefined && this.selectedProvider !== undefined) {
            this.selectedProvider([id]).subscribe(function (items) {
                items.forEach(function (item) {
                    var iqSelect2Item = _this.iqSelect2ItemAdapter(item);
                    _this.selectedItems = [iqSelect2Item];
                    _this.placeholderSelected = iqSelect2Item.text;
                });
            });
        }
    };
    IqSelect2Component.prototype.registerOnChange = function (fn) {
        this.onChangeCallback = fn;
    };
    IqSelect2Component.prototype.registerOnTouched = function (fn) {
        this.onTouchedCallback = fn;
    };
    IqSelect2Component.prototype.setDisabledState = function (isDisabled) {
        this.disabled = isDisabled;
    };
    IqSelect2Component.prototype.alreadySelected = function (item) {
        var result = false;
        this.selectedItems.forEach(function (selectedItem) {
            if (selectedItem.id === item.id) {
                result = true;
            }
        });
        return result;
    };
    IqSelect2Component.prototype.onItemSelected = function (item) {
        var _this = this;
        if (this.multiple) {
            this.selectedItems.push(item);
            var index = this.listData.indexOf(item, 0);
            if (index > -1) {
                this.listData.splice(index, 1);
            }
        }
        else {
            this.selectedItems.length = 0;
            this.selectedItems.push(item);
        }
        this.onChangeCallback('id' === this.referenceMode ? this.getSelectedIds() : this.getEntities());
        this.term.patchValue('', { emitEvent: false });
        setTimeout(function () { return _this.focusInput(); }, 1);
        this.resultsVisible = false;
        this.onSelect.emit(item);
        if (!this.multiple) {
            this.placeholderSelected = item.text;
        }
    };
    IqSelect2Component.prototype.getSelectedIds = function () {
        if (this.multiple) {
            var ids_1 = [];
            this.selectedItems.forEach(function (item) { return ids_1.push(item.id); });
            return ids_1;
        }
        else {
            return this.selectedItems.length === 0 ? null : this.selectedItems[0].id;
        }
    };
    IqSelect2Component.prototype.getEntities = function () {
        if (this.multiple) {
            var entities_1 = [];
            this.selectedItems.forEach(function (item) {
                entities_1.push(item.entity);
            });
            return entities_1;
        }
        else {
            return this.selectedItems.length === 0 ? null : this.selectedItems[0].entity;
        }
    };
    IqSelect2Component.prototype.removeItem = function (item) {
        var index = this.selectedItems.indexOf(item, 0);
        if (index > -1) {
            this.selectedItems.splice(index, 1);
        }
        this.onChangeCallback('id' === this.referenceMode ? this.getSelectedIds() : this.getEntities());
        this.onRemove.emit(item);
        if (!this.multiple) {
            this.placeholderSelected = '';
        }
    };
    IqSelect2Component.prototype.onFocus = function () {
        this.searchFocused = true;
    };
    IqSelect2Component.prototype.onBlur = function () {
        this.term.patchValue('', { emitEvent: false });
        this.searchFocused = false;
        this.resultsVisible = false;
        this.onTouchedCallback();
    };
    IqSelect2Component.prototype.getInputWidth = function () {
        var searchEmpty = this.selectedItems.length === 0 && (this.term.value === null || this.term.value.length === 0);
        var length = this.term.value === null ? 0 : this.term.value.length;
        if (!this.multiple) {
            return '100%';
        }
        else {
            return searchEmpty ? '100%' : (1 + length * .6) + 'em';
        }
    };
    IqSelect2Component.prototype.onKeyUp = function (ev) {
        if (this.results) {
            if (ev.keyCode === KEY_CODE_DOWN_ARROW) {
                this.results.activeNext();
            }
            else if (ev.keyCode === KEY_CODE_UP_ARROW) {
                this.results.activePrevious();
            }
            else if (ev.keyCode === KEY_CODE_ENTER) {
                this.results.selectCurrentItem();
            }
        }
        else {
            if (this.minimumInputLength === 0) {
                if (ev.keyCode === KEY_CODE_ENTER || ev.keyCode === KEY_CODE_DOWN_ARROW) {
                    this.focusInputAndShowResults();
                }
            }
        }
    };
    IqSelect2Component.prototype.onKeyDown = function (ev) {
        if (this.results) {
            if (ev.keyCode === KEY_CODE_TAB) {
                this.results.selectCurrentItem();
            }
        }
        if (ev.keyCode === KEY_CODE_DELETE) {
            if ((!this.term.value || this.term.value.length === 0) && this.selectedItems.length > 0) {
                this.removeItem(this.selectedItems[this.selectedItems.length - 1]);
            }
        }
    };
    IqSelect2Component.prototype.focusInput = function () {
        if (!this.disabled) {
            this.termInput.nativeElement.focus();
            this.resultsVisible = false;
        }
        this.searchFocused = !this.disabled;
    };
    IqSelect2Component.prototype.focusInputAndShowResults = function () {
        if (!this.disabled) {
            this.termInput.nativeElement.focus();
            this.subscribeToResults(Rx_1.Observable.of(''));
        }
        this.searchFocused = !this.disabled;
    };
    IqSelect2Component.prototype.onKeyPress = function (ev) {
        if (ev.keyCode === KEY_CODE_ENTER) {
            ev.preventDefault();
        }
    };
    IqSelect2Component.prototype.getCss = function () {
        return 'select2-selection-container ' + (this.css === undefined ? '' : this.css);
    };
    IqSelect2Component.prototype.getMinHeight = function () {
        var isInputSm = this.css === undefined ? false : this.css.indexOf('input-sm') !== -1;
        return isInputSm ? '30px' : '34px';
    };
    IqSelect2Component.prototype.getPlaceholder = function () {
        return this.selectedItems.length > 0 ? this.placeholderSelected : this.placeholder;
    };
    IqSelect2Component.prototype.isHideable = function () {
        return !this.multiple && this.placeholderSelected !== '';
    };
    IqSelect2Component.prototype.focus = function () {
        this.termInput.nativeElement.focus();
    };
    IqSelect2Component.prototype.getCountMessage = function () {
        var msg = this.messages && this.messages.moreResultsAvailableMsg ? this.messages.moreResultsAvailableMsg : this.MORE_RESULTS_MSG;
        msg = msg.replace(messages_1.Messages.PARTIAL_COUNT_VAR, String(this.listData.length));
        msg = msg.replace(messages_1.Messages.TOTAL_COUNT_VAR, String(this.resultsCount));
        return msg;
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Function)
    ], IqSelect2Component.prototype, "dataSourceProvider", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Function)
    ], IqSelect2Component.prototype, "selectedProvider", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Function)
    ], IqSelect2Component.prototype, "iqSelect2ItemAdapter", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], IqSelect2Component.prototype, "referenceMode", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], IqSelect2Component.prototype, "multiple", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], IqSelect2Component.prototype, "searchDelay", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], IqSelect2Component.prototype, "css", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], IqSelect2Component.prototype, "placeholder", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], IqSelect2Component.prototype, "minimumInputLength", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], IqSelect2Component.prototype, "disabled", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], IqSelect2Component.prototype, "searchIcon", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], IqSelect2Component.prototype, "deleteIcon", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', messages_1.Messages)
    ], IqSelect2Component.prototype, "messages", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], IqSelect2Component.prototype, "resultsCount", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], IqSelect2Component.prototype, "clientMode", void 0);
    __decorate([
        core_1.Output(), 
        __metadata('design:type', core_1.EventEmitter)
    ], IqSelect2Component.prototype, "onSelect", void 0);
    __decorate([
        core_1.Output(), 
        __metadata('design:type', core_1.EventEmitter)
    ], IqSelect2Component.prototype, "onRemove", void 0);
    __decorate([
        core_1.ViewChild('termInput'), 
        __metadata('design:type', Object)
    ], IqSelect2Component.prototype, "termInput", void 0);
    __decorate([
        core_1.ViewChild('results'), 
        __metadata('design:type', iq_select2_results_component_1.IqSelect2ResultsComponent)
    ], IqSelect2Component.prototype, "results", void 0);
    IqSelect2Component = __decorate([
        core_1.Component({
            selector: 'iq-select2',
            templateUrl: 'iq-select2.component.html',
            styleUrls: ['iq-select2.component.css'],
            providers: [VALUE_ACCESSOR]
        }), 
        __metadata('design:paramtypes', [])
    ], IqSelect2Component);
    return IqSelect2Component;
}());
exports.IqSelect2Component = IqSelect2Component;
